src/pyhf/constraints.py

```diff
@@ -38,85 +44,105 @@ def _precompute(self):
             except AttributeError:
                 normal_constraint_sigmas.append([1.0] * len(thisauxdata))
 
-            normal_constraint_data.append(thisauxdata)
-            normal_constraint_mean_indices.append(self.par_indices[parslice])
-
-        if normal_constraint_mean_indices:
-            normal_mean_idc = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='int'),
-                        normal_constraint_mean_indices,
-                    )
-                )
+        if self.parameter_helper.index_selection:
+            self._normal_data = default_backend.astensor(
+                default_backend.concatenate(normal_constraint_data), dtype='int'
             )
-            normal_sigmas = default_backend.concatenate(
-                list(map(default_backend.astensor, normal_constraint_sigmas))
+
+            _normal_sigmas = default_backend.concatenate(normal_constraint_sigmas)
+            sigmas = default_backend.reshape(_normal_sigmas, (1, -1))  # (1, normals)
+            self._batched_sigmas = default_backend.tile(
+                sigmas, (self.batch_size or 1, 1)
             )
-            normal_data = default_backend.concatenate(
-                list(
-                    map(
-                        lambda x: default_backend.astensor(x, dtype='int'),
-                        normal_constraint_data,
-                    )
+
+            # self.parameter_helper.index_selection  is [ (batch, parslice) ] ]
+            access_field = None
+            for x in self.parameter_helper.index_selection:
+                access_field = (
+                    x
+                    if access_field is None
+                    else default_backend.concatenate([access_field, x], axis=1)
                 )
-            )
+            # access field is (nbatch, normals)
+            self._access_field = access_field
 
-            self.normal_data = tensorlib.astensor(
-                default_backend.tolist(normal_data), dtype='int'
-            )
-            self.normal_sigmas = tensorlib.astensor(
-                default_backend.tolist(normal_sigmas)
-            )
-            self.normal_mean_idc = tensorlib.astensor(
-                default_backend.tolist(normal_mean_idc), dtype='int'
-            )
         else:
-            self.normal_data, self.normal_sigmas, self.normal_mean_idc = (
-                None,
-                None,
-                None,
-            )
+            self._normal_data = None
+            self._batched_sigmas = None
+            self._access_field = None
```

Comment:
set `self.batch_size = 0` here (instead of `None`) to remove the extra `if` check in `precompute`?

Reply:
I don't think introducing a magic number `self.batch_size = 0` to signal an unbatched computation is a good idea. I'd rather just keep the additional `if`

Reply:
Ok.